src/main/java/org/cryptomator/cloudaccess/requestdecorator/MetadataRequestDeduplicationDecorator.java

```diff
@@ -0,0 +1,55 @@
+package org.cryptomator.cloudaccess.requestdecorator;
+
+import com.github.benmanes.caffeine.cache.AsyncCache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import org.cryptomator.cloudaccess.api.CloudItemList;
+import org.cryptomator.cloudaccess.api.CloudItemMetadata;
+import org.cryptomator.cloudaccess.api.CloudPath;
+import org.cryptomator.cloudaccess.api.CloudProvider;
+
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * Decorates an existing CloudProvider by deduplicating identical itemMetadata and list-requests so that the delegate is called only once until the future is completed.
+ */
+class MetadataRequestDeduplicationDecorator extends CloudProviderDecorator {
+
+
+	// visible for testing
+	final AsyncCache<CloudPath, CloudItemMetadata> cachedItemMetadataRequests;
+	final AsyncCache<ItemListEntry, CloudItemList> cachedItemListRequests;
+
+	private final CloudProvider delegate;
+
+	public MetadataRequestDeduplicationDecorator(CloudProvider delegate) {
+		this(delegate, //
+				Caffeine.newBuilder().expireAfterWrite(Duration.ofSeconds(0)).buildAsync(), //
+				Caffeine.newBuilder().expireAfterWrite(Duration.ofSeconds(0)).buildAsync());
```

Comment:
Can you add a unit test that confirms that all requests resolve to the same CompletionStage until it gets resolved? Just to add some regression tests in case the behaviour changes.